src-electron/db/query-feature.js

```diff
@@ -24,53 +24,612 @@ const dbApi = require('./db-api.js')
 const dbMapping = require('./db-mapping.js')
 
 /**
- * Get all device type features associated with a list of device type refs
+ * Get all device type features associated with a list of device type refs and an endpoint.
+ * Join ENDPOINT_TYPE_ATTRIBUTE and ATTRIBUTE table to get featureMap attribute associated with the feature,
+ * so the frontend could get and set featureMap bit easier.
+ * Only return features with cluster on the side specified in the deivce type.
+ *
+ * @export
  * @param {*} db
  * @param {*} deviceTypeRefs
+ * @param {*} endpointTypeRef
  * @returns All feature information and device type conformance
- * with associated device type and cluster details
+ * with associated device type, cluster, and featureMap attribute details
  */
-async function getFeaturesByDeviceTypeRefs(db, deviceTypeRefs) {
+async function getFeaturesByDeviceTypeRefs(
+  db,
+  deviceTypeRefs,
+  endpointTypeRef
+) {
+  let arg = []
   let deviceTypeRefsSql = deviceTypeRefs.map(() => '?').join(', ')
+  arg.push(...deviceTypeRefs)
+  arg.push(endpointTypeRef)
   let features = await dbApi.dbAll(
     db,
     `
     SELECT
-        d.DESCRIPTION AS DEVICE_TYPE_NAME,
-        dc.CLUSTER_NAME,
-        dc.INCLUDE_SERVER,
-        dc.INCLUDE_CLIENT,
-        df.DEVICE_TYPE_CLUSTER_CONFORMANCE,
-        f.FEATURE_ID,
-        f.NAME AS FEATURE_NAME,
-        f.CODE,
-        f.BIT,
-        f.DEFAULT_VALUE,
-        f.DESCRIPTION
+			d.DESCRIPTION AS DEVICE_TYPE_NAME,
```

Comment:
Fix the formatting here.

Reply:
The format is in good shape if you click the actual file. Github is displaying it wrong weirdly